ykrt/src/mt.rs

```diff
@@ -578,16 +591,31 @@ impl MT {
                         HotLocationKind::Tracing => {
                             let hl = loc.hot_location_arc_clone().unwrap();
                             match &*mtt.tstate.borrow() {
-                                MTThreadState::Tracing { hl: thread_hl, .. } => {
+                                MTThreadState::Tracing {
+                                    hl: thread_hl,
+                                    frameaddr: tracing_frameaddr,
+                                    ..
+                                } => {
                                     // This thread is tracing something...
                                     if !Arc::ptr_eq(thread_hl, &hl) {
                                         // ...but not this Location.
                                         TransitionControlPoint::NoAction
                                     } else {
-                                        // ...and it's this location: we have therefore finished
-                                        // tracing the loop.
-                                        lk.kind = HotLocationKind::Compiling;
-                                        TransitionControlPoint::StopTracing
+                                        // ...and it's this location...
+                                        if frameaddr == *tracing_frameaddr {
+                                            lk.kind = HotLocationKind::Compiling;
+                                            TransitionControlPoint::StopTracing
+                                        } else {
+                                            // ...but in a different frame.
+                                            #[cfg(target_arch = "x86_64")]
```

Comment:
Nitpick: Here we could check for `StackDirection::GrowsDown` (defined in `regalloc.rs`).